README.md
# Composable Alternatives to Bevy's RunCriteria, States, FixedTimestep

This crate offers alternatives to the Run Criteria, States, and FixedTimestep
scheduling features currently offered by the Bevy game engine.

The ones provided by this crate do not use "looping stages", and can therefore
be combined/composed together elegantly, solving some of the most annoying
usability limitations of the respective APIs in Bevy.

Version Compatibility Table:

|Bevy Version|Crate Version      |
|------------|-------------------|
|`main`      |`bevy_main`        |
|`0.9`       |`main`             |
|`0.9`       |`0.9`              |
|`0.8`       |`0.7`, `0.8`       |
|`0.7`       |`0.4`, `0.5`, `0.6`|
|`0.6`       |`0.1`, `0.2`, `0.3`|

## How does this relate to the Bevy Stageless RFC?

This crate draws *very heavy* inspiration from the ["Stageless
RFC"](https://github.com/bevyengine/rfcs/pull/45) proposal for Bevy.

Big thanks to all the authors that have worked on that RFC and the designs
described there.

I am making this crate, because I believe the APIs currently in Bevy are
sorely in need of a usability improvement.

I figured out a way to implement the ideas from the Stageless RFC in a way
that works within the existing framework of current Bevy, without requiring
the complete scheduling API overhaul that the RFC proposes.

This way we can have something usable *now*, while the remaining Stageless
work is still in progress.

## Dependencies and Cargo Feature Flags

The "run conditions" functionality is always enabled, and depends only on
`bevy_ecs`.

The "fixed timestep" functionality is optional (`"fixedtimestep"` cargo
feature) and adds these dependencies:
 - `bevy_time`
 - `bevy_utils`

The "states" functionality is optional (`"states"` cargo feature) and adds
these dependencies:
 - `bevy_utils`

The `"app"` cargo feature enables extension traits that add new builder
methods to `App`, allowing more ergonomic access to the features of this
crate. Adds a dependency on `bevy_app`.

The `"bevy-compat"` feature adds Run Conditions for compatibility with
Bevy's legacy states implementation.

All of the optional cargo features are enabled by default.

## Run Conditions

This crate provides an alternative to Bevy Run Criteria, called "Run Conditions".

The different name was chosen to avoid naming conflicts and confusion with
the APIs in Bevy. Bevy Run Criteria are pretty deeply integrated into Bevy's
scheduling model, and this crate does not touch/replace them. They are
technically still there and usable.

### How Run Conditions Work?

You can convert any Bevy system into a "conditional system". This allows you
to add any number of "conditions" on it, by repeatedly calling the `.run_if`
builder method.

Each condition is just a Bevy system that outputs (returns) a `bool`.

The conditional system will present itself to
Bevy as a single big system (similar to Bevy's [system
piping](https://bevy-cheatbook.github.io/programming/system-chaining.html)),
combining the system it was created from with all the condition systems
attached.

When it runs, it will run each condition, and abort if any of them returns
`false`. The main system will run only if all the conditions return `true`.

(see `examples/conditions.rs` for a more complete example)

```rust
use bevy::prelude::*;
use iyes_loopless::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_system(
            notify_server
                .run_if(in_multiplayer)
                .run_if(on_mytimer)
        )
        .run();
}

/// Condition checking our timer
fn on_mytimer(mytimer: Res<MyTimer>) -> bool {
    mytimer.timer.just_finished()
}

/// Condition checking if we are connected to multiplayer server
fn in_multiplayer(gamemode: Res<GameMode>, connected: Res<ServerState>) -> bool {
    *gamemode == GameMode::Multiplayer &&
    connected.is_active()
}

/// Some system that should only run on a timer in multiplayer
fn notify_server(/* ... */) {
    // ...
}
```

It is highly recommended that all your condition systems only access data
immutably. Avoid mutable access or locals in condition systems, unless you are
really sure about what you are doing. If you add the same condition to many
systems, it *will run with each one*.

There are also some helper methods for easily adding common kinds of Run Conditions:
 - `.run_if_not`: invert the output of the condition
 - `.run_on_event::<T>()`: run if there are events of a given type
 - `.run_if_resource_exists::<T>()`: run if a resource of a given type exists
 - `.run_unless_resource_exists::<T>()`: run if a resource of a given type does not exist
 - `.run_if_resource_equals(value)`: run if the value of a resource equals the one provided
 - `.run_unless_resource_equals(value)`: run if the value of a resource does not equal the one provided

And if you are using [States](#states):
 - `.run_in_state(state)`
 - `.run_not_in_state(state)`

If you need to use classic Bevy States, you can use these adapters to check them with run conditions:
 - `.run_in_bevy_state(state)`
 - `.run_not_in_bevy_state(state)`

You can use Bevy labels for system ordering, as usual.

**Note:** conditional systems currently only support explicit labels, you cannot use
Bevy's "ordering by function name" syntax. E.g: `.after(another_system)` does *not* work,
you need to create a label.

There is also `ConditionSet` (similar to Bevy `SystemSet`): syntax sugar for
easily applying conditions and labels that are common to many systems:

```rust
use bevy::prelude::*;
use iyes_loopless::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_system(
            notify_server
                .run_if(in_multiplayer)
                .run_if(on_mytimer)
                // use bevy labels for ordering, as usual :)
                // (must be added at the end, after the conditions)
                .label("thing")
                .before("thing2")
        )
        // You can easily apply many conditions to many systems
        // using a `ConditionSet`:
        .add_system_set(ConditionSet::new()
            // all the conditions, and any labels/ordering
            // must be added before adding the systems
            // (helps avoid confusion and accidents)
            // (makes it clear they apply to all systems in the set)
            .run_if(in_multiplayer)
            .run_if(other_condition)
            .label("thing2")
            .after("stuff")
            .with_system(system1)
            .with_system(system2)
            .with_system(system3)
            .into() // Converts into Bevy `SystemSet` (to add to App)
        )
        .run();
}
```

**NOTE:** Due to some limitations with Bevy, `label`/`before`/`after` are
*not* supported on individual systems within a `ConditionSet`. You can only
use labels and ordering on the entire set, to apply them to all member
systems. If some systems need different ordering, just add them individually
with `.add_system`.

## Fixed Timestep

This crate offers a fixed timestep implementation that runs as a separate
Stage in the Bevy schedule. This way, it does not conflict with any other
functionality. You can easily use [run conditions](#run-conditions) and
[states](#states) to control your fixed timestep systems.

It is possible to add multiple "sub-stages" within a fixed timestep, allowing
you to apply `Commands` within a single timestep run. For example, if you want
to spawn entities and then do something with them, on the same tick.

It is also possible to have multiple independent fixed timesteps, should you
need to.

(see `examples/fixedtimestep.rs` for a more complex working example)

```rust
use bevy::prelude::*;
use iyes_loopless::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        // add the fixed timestep stage:
        // (in the default position, before CoreStage::Update)
        .add_fixed_timestep(
            Duration::from_millis(250),
            // we need to give it a string name, to refer to it
            "my_fixed_update",
        )
        // add fixed timestep systems:
        .add_fixed_timestep_system(
            "my_fixed_update", 0, // fixed timestep name, sub-stage index
            // it can be a conditional system!
            my_simulation
                .run_if(some_condition)
                .run_in_state(AppState::InGame)
                .after("some_label")
        )
        .run();
}
```

Every frame, the `FixedTimestepStage` will accumulate the time delta. When
it goes over the set timestep value, it will run all the child stages. It
will repeat the sequence of child stages multiple times if needed, if
more than one timestep has accumulated.

### Fixed Timestep Control

You can use the `FixedTimesteps` resource (make sure it is the one from this
crate, not the one from Bevy with the same name) to access information about a
fixed timestep and to control its parameters, like the timestep duration.

```rust
fn timestep_control(mut timesteps: ResMut<FixedTimestep>) {
    // we can access our timestep by name
    let info = timesteps.get_mut("my_fixed_update").unwrap();
    // set a different duration
    info.step = Duration::from_millis(125);
    // pause it
    info.paused = true;
}

/// Print info about the fixed timestep this system runs in
fn debug_fixed(timesteps: Res<FixedTimesteps>) {
    // from within a system that runs inside the fixed timestep,
    // you can use `.get_current`, no need for the timestep name:
    let info = timesteps.get_current().unwrap();
    println!("Fixed timestep duration: {:?} ({} Hz).", info.timestep(), info.rate());
    println!("Overstepped by {:?} ({}%).", info.remaining(), info.overstep() * 100.0);
}
```

## States

(see `examples/menu.rs` for a complete example)

This crate offers a states abstraction that works as follows:

You create one (or more) state types, usually enums, just like when using
Bevy States.

However, here we track states using two resource types:
 - `CurrentState(T)`: the current state you are in
 - `NextState(T)`: insert this (using `Commands`) whenever you want to change state

### Registering the state type

You need to add the state to your `App` using `.add_loopless_state(value)`
with the initial state value. This helper method adds a special stage type
(`StateTransitionStage`) that is responsible for performing state transitions.
By default, it is added before `CoreStage::Update`. If you would like the
transitions to be executed elsewhere in the app schedule, there are other
helper methods that let you specify the position.

For advanced use cases, you could construct and add the `StateTransitionStage`
manually, without the helper method.

### Enter/Exit Systems

You can add enter/exit systems to be executed on state transitions, using
`.add_enter_system(state, system)` and `.add_exit_system(state, system)`.

For advanced scenarios, you could add a custom stage type instead, using
`.set_enter_stage(state, stage)` and `.set_exit_stage(state, stage)`.

### State Transition

When the `StateTransitionStage` runs, it will check if a `NextState` resource
exists. If yes, it will remove it and perform a transition:
 - run the "exit stage" (if any) for the current state
 - change the value of `CurrentState`
 - run the "enter stage" (if any) for the next state

If you want to perform a state transition, simply insert a `NextState<T>`.
If you mutate `CurrentState<T>`, you will effectively change state without
running the exit/enter systems (you probably don't want to do this).

Multiple state transitions can be performed in a single frame, if you insert
a new instance of `NextState` from within an exit/enter stage.

### Update systems

For the systems that you want to run every frame, we provide
a `.run_in_state(state)` and `.run_not_in_state(state)` [run
conditions](#run-conditions).

You can add systems anywhere, to any stage (incl. behind [fixed
timestep](#fixed-timestep)), and make them conditional on one or more states,
using those helper methods.

```rust
use bevy::prelude::*;
use iyes_loopless::prelude::*;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
enum GameState {
    MainMenu,
    InGame,
}

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        // Add our state type
        .add_loopless_state(GameState::MainMenu)
        // If we had more state types, we would add them too...

        // Add a FixedTimestep, cuz we can!
        .add_fixed_timestep(
            Duration::from_millis(250),
            "my_fixed_update",
        )
        .add_fixed_timestep_system(
            "my_fixed_update", 0,
            my_simulation
                .run_in_state(AppState::InGame)
        )

        // Add our various systems
        .add_system(menu_stuff.run_in_state(GameState::MainMenu))
        .add_system(animate.run_in_state(GameState::InGame))

        // On states Enter and Exit
        .add_enter_system(GameState::MainMenu, setup_menu)
        .add_exit_system(GameState::MainMenu, despawn_menu)
        .add_enter_system(GameState::InGame, setup_game)

        .run();
}

```

### State transitions under fixed timestep

If you have a state type that you are using for controlling fixed timestep
stuff, you might want state transitions to happen only on fixed timestep
(not just on any frame).

To accomplish that, you can add the `StateTransitionStage` as a child stage
at the beginning of your `FixedTimestepStage`.

The stage types from this crate are composable like that! :) They accept
any stage type.


CHANGELOG.md
# Changelog

Notable user-facing changes with each release version will be described in this file.

## [0.9.1]: 2022-11-20

### Fixed
 - Conditional exclusive systems no longer panic when ran

## [0.9.0]: 2022-11-19

### Changed
 - Bevy 0.9 compatibility

## [0.8.0]: 2022-10-24

### Added
 - All missing API docs
 - Extension traits for a nice API on App/Schedule for working with fixed timesteps, similar to states
 - `FixedTimesteps` resource: allows any system to access the properties of any fixed timestep
 - Fixed timesteps can be paused/unpaused

### Changed
 - `FixedTimestepInfo` is now accessed via a `FixedTimesteps` resource
 - Fixed timestep APIs use string names to identify fixed timesteps
 - Create a conditional exclusive system by calling `.into_conditional_exclusive()`.
   No more conflicting traits. No need for special imports, prelude just works.

### Removed
 - `FixedTimestepInfo` is no longer directly provided as a resource

### Fixed
 - `run_on_event` run condition no longer fires twice under some edge cases
 - WASM compatibility for fixed timestep: use `bevy_utils::Duration` instead of `std::time::Duration`

## [0.7.1]: 2022-08-18

### Added
 - Optional `bevy-inspector-egui` support (thanks @jakobhellermann)

### Changed
 - Using bare system function names with `before`/`after` is now a compile error instead of runtime warning.
   (this was always broken and unsupported)

## [0.7.0]: 2022-07-31

### Added
 - API helper extension methods for `Schedule`, analogous to those for `App`. (thanks @NiklasEi)

### Changed
 - Bevy 0.8 support
 - `FixedTimestepInfo.accumulator` is now `pub`; mutations also affect the internal accumulator

## [0.6.1]: 2022-06-20

### Changed
 - The `step` field in `FixedTimestepInfo` is now `pub`. This was a mistake in 0.6.0.

## [0.6.0]: 2022-06-15

### Added
 - `add_{enter,exit}_system_set` helpers for adding multiple systems to the enter/exit stages of states.
 - `run_if_resource_added` and `run_if_resource_removed` run conditions (thanks @Shatur)

### Changed
 - It is now possible to reconfigure fixed timestep durations at runtime, by modifying the `step` field in `FixedTimestepInfo`.

## [0.5.1]: 2022-04-24

### Added
 - Support for labels/ordering on `ConditionSet`

## [0.5.0]: 2022-04-22

### Added
 - Support for conditional exclusive systems, using the `IntoConditionalExclusiveSystem` trait

### Changed
 - `.add_{enter,exit}_system` App helpers no longer use a `&` reference to the state
 - The `.run_if*` methods are now in trait `ConditionHelpers`, not inherent on the type

## [0.4.0]: 2022-04-16

### Added
 - Extension trait to add ergonomic helpers to `App` for using states.
   - (optional behind `app` feature, adds `bevy_app` dependency)

### Changed
 - Updated for Bevy 0.7

## [0.3.0]: 2022-04-13

### Changed
 - Reverted the `NextState` behavior to how it was in `0.1.x`. The resource has to be inserted/removed.
   In retrospect, this is better UX and avoids bugs.
   - However, support transitioning to the same state as the current.

## [0.2.1]: 2022-04-11

### Added

 - Fixed Timestep: optional EXPERIMENTAL "rate lock" algorithm (see api docs)

## [0.2.0]: 2022-04-06

### Added

 - `ConditionSet`: makes it easy to add run conditions to many systems at once.
 - `FixedTimestepInfo` resource: allows your fixed timestep systems to know about the parameters of the current fixed timestep.

### Changed
 - Behavior of `NextState`: Checked using Bevy Change Detection
   - Present at all times, not removed on state transition.
   - No longer required to be inserted using `Commands`; you can also mutate it directly. Either way works.
   - Supports "transitioning" to the same state as the current, to "reset" it.
 - Conditional systems are now boxed, not generic.

### Removed
 - Conditional systems no longer support `In` and `Out` parameters.

## [0.1.1]: 2022-03-23

### Added
 - Run Condition adapters for compatibility with legacy Bevy States (`.run_in_bevy_state()`/`.run_not_in_bevy_state()`)

### Changed
 - Manually calling `.into_conditional()` on systems, to add conditions, is no longer required.

## [0.1.0]: 2022-03-21

Initial Release

[0.9.1]: https://github.com/IyesGames/iyes_loopless/tree/v0.9.1
[0.9.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.9.0
[0.8.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.8.0
[0.7.1]: https://github.com/IyesGames/iyes_loopless/tree/v0.7.1
[0.7.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.7.0
[0.6.1]: https://github.com/IyesGames/iyes_loopless/tree/v0.6.1
[0.6.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.6.0
[0.5.1]: https://github.com/IyesGames/iyes_loopless/tree/v0.5.1
[0.5.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.5.0
[0.4.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.4.0
[0.3.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.3.0
[0.2.1]: https://github.com/IyesGames/iyes_loopless/tree/v0.2.1
[0.2.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.2.0
[0.1.1]: https://github.com/IyesGames/iyes_loopless/tree/v0.1.1
[0.1.0]: https://github.com/IyesGames/iyes_loopless/tree/v0.1.0


src/lib.rs
//! Composable Alternatives to Bevy's RunCriteria, States, FixedTimestep
//! 
//! This crate offers alternatives to the Run Criteria, States, and FixedTimestep
//! scheduling features currently offered by the Bevy game engine.
//! 
//! The ones provided by this crate do not use "looping stages", and can therefore
//! be combined/composed together elegantly, solving some of the most annoying
//! usability limitations of the respective APIs in Bevy.

#![warn(missing_docs)]

pub mod condition;
#[cfg(feature = "fixedtimestep")]
pub mod fixedtimestep;
#[cfg(feature = "states")]
pub mod state;

/// Prelude: convenient import for all the user-facing APIs provided by the crate
pub mod prelude {
    pub use crate::condition::{ConditionHelpers, IntoConditionalSystem, ConditionSet, AddConditionalToSet};

    #[cfg(feature = "fixedtimestep")]
    pub use crate::fixedtimestep::{FixedTimesteps, FixedTimestepStage};
    #[cfg(feature = "fixedtimestep")]
    pub use crate::fixedtimestep::schedule::ScheduleLooplessFixedTimestepExt;
    #[cfg(all(feature = "fixedtimestep", feature = "app"))]
    pub use crate::fixedtimestep::app::AppLooplessFixedTimestepExt;

    #[cfg(feature = "states")]
    pub use crate::state::{CurrentState, NextState, StateTransitionStage};
    #[cfg(feature = "states")]
    pub use crate::state::schedule::ScheduleLooplessStateExt;
    #[cfg(all(feature = "states", feature = "app"))]
    pub use crate::state::app::AppLooplessStateExt;
}


examples/menu.rs
//! Complex example showcasing all the features together.
//!
//! Shows how our states, fixed timestep, and custom run conditions, can all be used together!
//!
//! Also shows how run conditions could be helpful for Bevy UI button handling!
//!
//! This example has a main menu with two buttons: exit the app and enter the game.
//!
//! How to "play the game": hold spacebar to spawn colorful squares, release spacebar to make them spin! <3

use bevy::prelude::*;
use iyes_loopless::prelude::*;

use bevy::app::AppExit;
use bevy::window::close_on_esc;

use std::time::Duration;

use rand::prelude::*;

/// Our Application State
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
enum GameState {
    MainMenu,
    InGame,
}

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        // add out states driver
        .add_loopless_state(GameState::MainMenu)
        // Add a FixedTimestep, cuz we can!
        .add_fixed_timestep(
            Duration::from_millis(125),
            // give it a label
            "my_fixed_update",
        )
        // menu setup (state enter) systems
        .add_enter_system(GameState::MainMenu, setup_menu)
        // menu cleanup (state exit) systems
        .add_exit_system(GameState::MainMenu, despawn_with::<MainMenu>)
        // game cleanup (state exit) systems
        .add_exit_system(GameState::InGame, despawn_with::<MySprite>)
        // menu stuff
        .add_system_set(
            ConditionSet::new()
                .run_in_state(GameState::MainMenu)
                .with_system(close_on_esc)
                .with_system(butt_interact_visual)
                // our menu button handlers
                .with_system(butt_exit.run_if(on_butt_interact::<ExitButt>))
                .with_system(butt_game.run_if(on_butt_interact::<EnterButt>))
                .into()
        )
        // in-game stuff
        .add_system_set(
            ConditionSet::new()
                .run_in_state(GameState::InGame)
                .with_system(back_to_menu_on_esc)
                .with_system(clear_on_del)
                .with_system(spin_sprites.run_if_not(spacebar_pressed))
                .into()
        )
        .add_fixed_timestep_system(
            "my_fixed_update", 0,
            spawn_sprite
                // only in-game!
                .run_in_state(GameState::InGame)
                // only while the spacebar is pressed
                .run_if(spacebar_pressed)
        )
        // our other various systems:
        .add_system(debug_current_state)
        // setup our camera globally (for UI) at startup and keep it alive at all times
        .add_startup_system(setup_camera)
        .run();
}

/// Marker for our in-game sprites
#[derive(Component)]
struct MySprite;

/// Marker for the main menu entity
#[derive(Component)]
struct MainMenu;

/// Marker for the main game camera entity
#[derive(Component)]
struct GameCamera;

/// Marker for the "Exit App" button
#[derive(Component)]
struct ExitButt;

/// Marker for the "Enter Game" button
#[derive(Component)]
struct EnterButt;

/// Reset the in-game state when pressing delete
fn clear_on_del(mut commands: Commands, kbd: Res<Input<KeyCode>>) {
    if kbd.just_pressed(KeyCode::Delete) || kbd.just_pressed(KeyCode::Back) {
        commands.insert_resource(NextState(GameState::InGame));
    }
}

/// Transition back to menu on pressing Escape
fn back_to_menu_on_esc(mut commands: Commands, kbd: Res<Input<KeyCode>>) {
    if kbd.just_pressed(KeyCode::Escape) {
        commands.insert_resource(NextState(GameState::MainMenu));
    }
}

/// We can just access the `CurrentState`, and even use change detection!
fn debug_current_state(state: Res<CurrentState<GameState>>) {
    if state.is_changed() {
        println!("Detected state change to {:?}!", state);
    }
}

/// Condition system for holding the space bar
fn spacebar_pressed(kbd: Res<Input<KeyCode>>) -> bool {
    kbd.pressed(KeyCode::Space)
}

/// Despawn all entities with a given component type
fn despawn_with<T: Component>(mut commands: Commands, q: Query<Entity, With<T>>) {
    for e in q.iter() {
        commands.entity(e).despawn_recursive();
    }
}

/// Spawn a MySprite entity
fn spawn_sprite(mut commands: Commands) {
    let mut rng = thread_rng();
    commands.spawn((SpriteBundle {
        sprite: Sprite {
            color: Color::rgba(rng.gen(), rng.gen(), rng.gen(), 0.5),
            custom_size: Some(Vec2::new(64., 64.)),
            ..Default::default()
        },
        transform: Transform::from_xyz(
            rng.gen_range(-420.0..420.0),
            rng.gen_range(-420.0..420.0),
            rng.gen_range(0.0..100.0),
        ),
        ..Default::default()
    }, MySprite));
}

/// Spawn the camera
fn setup_camera(mut commands: Commands) {
    commands.spawn((Camera2dBundle::default(), GameCamera));
}

/// Rotate all the sprites
fn spin_sprites(mut q: Query<&mut Transform, With<MySprite>>, t: Res<Time>) {
    for mut transform in q.iter_mut() {
        transform.rotate(Quat::from_rotation_z(1.0 * t.delta_seconds()));
    }
}

/// Change button color on interaction
fn butt_interact_visual(
    mut query: Query<(&Interaction, &mut BackgroundColor), (Changed<Interaction>, With<Button>)>,
) {
    for (interaction, mut color) in query.iter_mut() {
        match interaction {
            Interaction::Clicked => {
                *color = BackgroundColor(Color::rgb(0.75, 0.75, 0.75));
            }
            Interaction::Hovered => {
                *color = BackgroundColor(Color::rgb(0.8, 0.8, 0.8));
            }
            Interaction::None => {
                *color = BackgroundColor(Color::rgb(1.0, 1.0, 1.0));
            }
        }
    }
}

/// Condition to help with handling multiple buttons
///
/// Returns true when a button identified by a given component is clicked.
fn on_butt_interact<B: Component>(
    query: Query<&Interaction, (Changed<Interaction>, With<Button>, With<B>)>,
) -> bool {
    for interaction in query.iter() {
        if *interaction == Interaction::Clicked {
            return true;
        }
    }

    false
}

/// Handler for the Exit Game button
fn butt_exit(mut ev: EventWriter<AppExit>) {
    ev.send(AppExit);
}

/// Handler for the Enter Game button
fn butt_game(mut commands: Commands) {
    // queue state transition
    commands.insert_resource(NextState(GameState::InGame));
}

/// Construct the main menu UI
fn setup_menu(mut commands: Commands, ass: Res<AssetServer>) {
    let butt_style = Style {
        justify_content: JustifyContent::Center,
        align_items: AlignItems::Center,
        padding: UiRect::all(Val::Px(8.0)),
        margin: UiRect::all(Val::Px(4.0)),
        flex_grow: 1.0,
        ..Default::default()
    };
    let butt_textstyle = TextStyle {
        font: ass.load("Sansation-Regular.ttf"),
        font_size: 24.0,
        color: Color::BLACK,
    };

    let menu = commands
        .spawn((NodeBundle {
            background_color: BackgroundColor(Color::rgb(0.5, 0.5, 0.5)),
            style: Style {
                size: Size::new(Val::Auto, Val::Auto),
                margin: UiRect::all(Val::Auto),
                align_self: AlignSelf::Center,
                flex_direction: FlexDirection::Column,
                justify_content: JustifyContent::Center,
                ..Default::default()
            },
            ..Default::default()
        }, MainMenu))
        .id();

    let butt_enter = commands
        .spawn((ButtonBundle {
            style: butt_style.clone(),
            ..Default::default()
        }, EnterButt))
        .with_children(|btn| {
            btn.spawn(TextBundle {
                text: Text::from_section("Enter Game", butt_textstyle.clone()),
                ..Default::default()
            });
        })
        .id();

    let butt_exit = commands
        .spawn((ButtonBundle {
            style: butt_style.clone(),
            ..Default::default()
        }, ExitButt))
        .with_children(|btn| {
            btn.spawn(TextBundle {
                text: Text::from_section("Exit Game", butt_textstyle.clone()),
                ..Default::default()
            });
        })
        .id();

    commands
        .entity(menu)
        .push_children(&[butt_enter, butt_exit]);
}


examples/conditions.rs
use bevy::prelude::*;
use iyes_loopless::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .insert_resource(MyTimer::new())
        .insert_resource(GameMode::Multiplayer)
        .init_resource::<ServerState>()
        .add_system(
            notify
                .run_if(in_multiplayer)
                .run_if(on_mytimer)
                // labels and ordering must come at the end
                .after("tick"),
        )
        .add_system(
            tick_mytimer
                .run_if(in_multiplayer)
                .run_if(spacebar_pressed)
                .label("tick")
        )
        .run();
}

/// Condition checking our timer
fn on_mytimer(mytimer: Res<MyTimer>) -> bool {
    mytimer.timer.just_finished()
}

/// Condition checking if we are connected to multiplayer server
fn in_multiplayer(gamemode: Res<GameMode>, connected: Res<ServerState>) -> bool {
    *gamemode == GameMode::Multiplayer && connected.is_active()
}

/// Condition checking if spacebar is pressed
fn spacebar_pressed(kbd: Res<Input<KeyCode>>) -> bool {
    kbd.pressed(KeyCode::Space)
}

fn notify() {
    println!("BAM!");
}

/// Timers gotta be ticked
fn tick_mytimer(mut mytimer: ResMut<MyTimer>, time: Res<Time>) {
    mytimer.timer.tick(time.delta());
}

#[derive(Resource)]
struct MyTimer {
    timer: Timer,
}

impl MyTimer {
    fn new() -> Self {
        Self {
            timer: Timer::from_seconds(1.0, TimerMode::Repeating),
        }
    }
}

#[derive(PartialEq, Eq)]
#[derive(Resource)]
pub enum GameMode {
    Singleplayer,
    Multiplayer,
}

#[derive(Default)]
#[derive(Resource)]
struct ServerState {
    // ...
}

impl ServerState {
    fn is_active(&self) -> bool {
        // placeholder
        true
    }
}


examples/fixedtimestep.rs
use bevy::prelude::*;
use iyes_loopless::prelude::*;
use rand::prelude::*;

use std::time::Duration;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)

        // add fixed timestep stage to the default location (before Update)
        .add_fixed_timestep(
            Duration::from_millis(250),
            // give it a label
            "my_fixed_update",
        )

        // add an additional child "sub-stage" under the fixed timestep;
        // this will let us apply Commands within one fixed timestep run
        .add_fixed_timestep_child_stage("my_fixed_update")

        // add a system to our fixed timestep (first sub-stage)
        .add_fixed_timestep_system("my_fixed_update", 0, debug_fixed_timestep)

        // to showcase use of Commands, we will spawn entities in one sub-stage (0) ...
        .add_fixed_timestep_system("my_fixed_update", 0, spawn_entities)
        // ... and mutate their transform in another (1)
        .add_fixed_timestep_system("my_fixed_update", 1, reposition_entities)

        .add_startup_system(setup_camera)
        .add_system(debug_new_count)
        .add_system(random_hiccups)
        .add_system(kbd_control_timestep)
        .add_system(clear_entities)
        .run();
}

#[derive(Component)]
struct MySprite;

/// Spawn a MySprite entity
fn spawn_entities(mut commands: Commands) {
    let mut rng = thread_rng();

    commands.spawn((SpriteBundle {
        sprite: Sprite {
            color: Color::rgba(rng.gen(), rng.gen(), rng.gen(), 0.5),
            custom_size: Some(Vec2::new(64., 64.)),
            ..Default::default()
        },
        // the `reposition_entities` system will take care of X and Y ;)
        transform: Transform::from_xyz(0.0, 0.0, rng.gen_range(0.0..100.0)),
        ..Default::default()
    }, MySprite));
}

/// Move each sprite to a random X,Y position
fn reposition_entities(mut q: Query<&mut Transform, With<MySprite>>) {
    let mut rng = thread_rng();

    for mut transform in q.iter_mut() {
        transform.translation.x = rng.gen_range(-420.0..420.0);
        transform.translation.y = rng.gen_range(-420.0..420.0);
    }
}

/// Every fixed timestep, print info about the timestep parameters
/// (shows how to get it from FixedTimesteps)
fn debug_fixed_timestep(timesteps: Res<FixedTimesteps>) {
    // unwrap: this system will run inside of the fixed timestep
    let info = timesteps.get_current().unwrap();
    println!("Fixed timestep duration: {:?} ({} Hz).", info.timestep(), info.rate());
    println!("Overstepped by {:.2?} ({:.2}%).", info.remaining(), info.overstep() * 100.0);
}

/// Every frame, print if new MySprites have been spawned
fn debug_new_count(q: Query<(), Added<MySprite>>) {
    let new = q.iter().count();
    if new > 0 {
        println!("{:?} new sprites spawned this frame", new);
        println!();
    }
}

/// Randomly decide to sleep for a while to simulate "lag spikes"
///
/// Used to showcase fixed timestep behavior in such situations
fn random_hiccups() {
    let mut rng = rand::thread_rng();

    if rng.gen::<u8>() == 0 {
        std::thread::sleep(Duration::from_millis(500));
    }

    if rng.gen::<u8>() == 255 {
        std::thread::sleep(Duration::from_millis(1000));
    }
}

/// Keypresses for speeding up / slowing down / pausing the fixed timestep
/// (by mutating the FixedTimestepInfo from FixedTimesteps)
fn kbd_control_timestep(
    kbd: Res<Input<KeyCode>>,
    mut timesteps: ResMut<FixedTimesteps>,
) {
    // this system runs outside of the fixed timestep, so we need
    // to get the fixed timestep info by label
    let info = timesteps.get_mut("my_fixed_update").unwrap();

    if kbd.any_just_pressed([KeyCode::Minus, KeyCode::Underline]) {
        info.step = Duration::from_secs_f32(info.step.as_secs_f32() * 0.75);
    }
    if kbd.any_just_pressed([KeyCode::Plus, KeyCode::Equals]) {
        info.step = Duration::from_secs_f32(info.step.as_secs_f32() * 1.25);
    }
    if kbd.just_pressed(KeyCode::Space) {
        info.toggle_pause();
    }
}

/// Clear entities with keypress
fn clear_entities(
    mut commands: Commands,
    kbd: Res<Input<KeyCode>>,
    q: Query<Entity, With<MySprite>>
) {
    if kbd.any_just_pressed([KeyCode::Delete, KeyCode::Back]) {
        for e in q.iter() {
            commands.entity(e).despawn();
        }
    }
}

fn setup_camera(mut commands: Commands) {
    commands.spawn(Camera2dBundle::default());
}
